### PR TITLE
Implement data exporting

### PR DESCRIPTION
**Pull request by s3364859**
In this PR:
- Added in the ability to export a dataset as a CSV. This required moving a fair bit of stuff around
- Consolidated a lot of implementation, especially in the object model
- Please do not add code to this PR

export

### CODE REVIEW

File: CPT331.Core/ObjectModel/Crime.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ namespace CPT331.Core.ObjectModel
 	/// <summary>
 	/// Represents a Crime type, used to describe information about a crime.
 	/// </summary>
-	public class Crime
+	public class Crime : ReadOnlyDataObject
```
Comment (s3364859) [inline CPT331.Core/ObjectModel/Crime.cs]: Needed a type to group exportable object models by.

File: CPT331.Core/ObjectModel/ReadOnlyDataObject.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+using System;
+
+#endregion
+
+namespace CPT331.Core.ObjectModel
+{
+	/// <summary>
+	/// Represents a ReadOnlyDataObject type, which contains immutable information obtained fron a data source.
+	/// </summary>
+	public class ReadOnlyDataObject
```
Comment (s3364859) [inline CPT331.Core/ObjectModel/ReadOnlyDataObject.cs]: This is the base object model type (atm anyway). Allows for many object model types to be treated the same way.

File: CPT331.Data/AdhocScriptRepository.cs [hunk 3]
```diff
@@ -15,50 +15,7 @@ namespace CPT331.Data
 	/// <summary>
 	/// Represents an AdhocScriptRepository type, used to execute SQL against a data source.
 	/// </summary>
-	public static class AdhocScriptRepository
+	public class AdhocScriptRepository : Repository
```
Comment (s3364859) [inline CPT331.Data/AdhocScriptRepository.cs]: Also needed a way to be able to treat all repositories the same way.

File: CPT331.Data/DataProvider.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+using System;
+
+#endregion
+
+namespace CPT331.Data
+{
+	/// <summary>
+	/// Represents a DataProvider type, used as a repository factory.
+	/// </summary>
+	public static class DataProvider
```
Comment (s3364859) [inline CPT331.Data/DataProvider.cs]: Factory type, used to manage repository objects.

File: CPT331.Data/OffenceRepository.cs [hunk 5]
```diff
@@ -100,15 +101,27 @@ public static List<Offence> GetOffences()
...
-        /// <param name="id">ID indicating the Offence record to be updated</param>
-        /// <param name="isDeleted">New value for the IsDeleted column</param>
-        /// <param name="isVisible">New value for the IsVisible column</param>
-        /// <param name="name">New value for the Name column</param>
+		/// <summary>
+		/// Exports a list of ReadOnlyDataObject types to a stream.
+		/// </summary>
+		/// <param name="readOnlyDataObjects">The list of ReadOnlyDataObject objects to export.</param>
+		/// <param name="stream">The stream to export the ReadOnlyDataObject list to.</param>
+		protected override void OnExport(List<ReadOnlyDataObject> readOnlyDataObjects, Stream stream)
```
Comment (s3364859) [inline CPT331.Data/OffenceRepository.cs]: This is where a list of objects this repository creates is pushed to the base for handling.

File: CPT331.Data/Repository.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+using System.Reflection;
+
+#endregion
+
+namespace CPT331.Data
+{
+	/// <summary>
+	/// Represents a Repository type, used to manipulate data.
+	/// </summary>
+	public class Repository
```
Comment (s3364859) [inline CPT331.Data/Repository.cs]: Base (for now) repository implementation.

File: CPT331.Data/Repository.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+			return executeScript;
+		}
+
+		/// <summary>
+		/// Begins an export operation to a stream.
+		/// </summary>
+		/// <param name="stream">The stream to export to.</param>
+		public void Export(Stream stream)
+		{
```
Comment (s3364859) [inline CPT331.Data/Repository.cs]: New method to export data to a stream.

File: CPT331.Data/Repository.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		}
+
+		/// <summary>
+		/// Exports a list of ReadOnlyDataObject types to a stream.
+		/// </summary>
+		/// <param name="readOnlyDataObjects">The list of ReadOnlyDataObject objects to export.</param>
+		/// <param name="stream">The stream to export the ReadOnlyDataObject list to.</param>
+		protected virtual void OnExport(List<ReadOnlyDataObject> readOnlyDataObjects, Stream stream)
+		{
+			if ((readOnlyDataObjects != null) && (readOnlyDataObjects.Count > 0))
```
Comment (s3364859) [inline CPT331.Data/Repository.cs]: Here's where the heavy lifting is done.

File: CPT331.Data/Repository.cs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+					ReadOnlyDataObject headerReadOnlyDataObject = readOnlyDataObjects[0];
+					List<PropertyInfo> propertyInfos = headerReadOnlyDataObject.GetType().GetProperties().ToList<PropertyInfo>();
+
+					//	Header
+					propertyInfos.ForEach(m => streamWriter.Write($"{m.Name},"));
+					streamWriter.WriteLine();
+
+					// Rows
+
+					foreach (ReadOnlyDataObject readOnlyDataObject in readOnlyDataObjects)
```
Comment (s3364859) [inline CPT331.Data/Repository.cs]: Yes, reflection is slow, I know /uppercut

File: CPT331.Web/Actions/ExportDataActionResult.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+using System;
+using System.Web.Mvc;
+
+using CPT331.Data;
+
+#endregion
+
+namespace CPT331.Web.Actions
+{
+	public class ExportDataActionResult : ActionResult
```
Comment (s3364859) [inline CPT331.Web/Actions/ExportDataActionResult.cs]: We can now return a new type of action result from our controllers - this prompts the user to download a CSV.

File: CPT331.Web/Controllers/AdminController.cs [hunk 11]
```diff
@@ -9,44 +9,44 @@
...
-
-        #region Crime Model
-        /// <summary>
-        /// Provides an editor page for an existing Crime record.
-        /// </summary>
-        /// <param name="id">The unique ID for the Crime record.</param>
-        /// <returns>The 'Edit Crime' view.</returns>
-        /// <permission cref="AdminAuthorizeAttribute">Only authorised personnel have access to this method.</permission>
-        [HttpGet]
+	/// <summary>
```
Comment (s3364859) [inline CPT331.Web/Controllers/AdminController.cs]: One messy merge due to the reordering of methods and such. It should be all in there though.

File: CPT331.Web/Controllers/AdminController.cs [hunk 12]
```diff
@@ -191,54 +190,61 @@ public ActionResult Crimes(string sortBy, SortDirection? sortDirection, uint? pa
...
-        /// <param name="id">The unique ID for the LocalGovernmentArea record.</param>
-        /// <returns>The 'Edit LocalGovernmentArea' view.</returns>
-        /// <permission cref="AdminAuthorizeAttribute">Only authorised personnel have access to this method.</permission>
-        [HttpGet]
-        public ActionResult LocalGovernmentArea(uint id)
+		#endregion
+
+		#region LocalGovernmentArea Model
+		[HttpGet]
+		public ActionResult ExportLocalGovernmentAreas()
```
Comment (s3364859) [inline CPT331.Web/Controllers/AdminController.cs]: Export action - uncommented at this stage.

File: CPT331.Web/Views/Admin/LocalGovernmentAreas.cshtml [hunk 13]
```diff
@@ -7,11 +7,14 @@
...
 }
 
-<br/>
+<br />
 <h1>@Html.ActionLink("Home", "Home", "Admin") | Local Government Areas</h1>
 <hr /><br />
 
 <p><h3>@Html.ActionLink("New Local Government Area", "NewLocalGovernmentArea", "Admin")</h3></p>
+
+<p><a href="@Url.Action("ExportLocalGovernmentAreas", "Admin")"><img src="/Content/Images/excel.gif" border="0" title="Export Data" /></a></p>
```
Comment (s3364859) [inline CPT331.Web/Views/Admin/LocalGovernmentAreas.cshtml]: Simple addition to the views to allow the user to download the corresponding data.

File: CPT331.Web/Web.config [hunk 14]
```diff
@@ -32,6 +32,7 @@
 	</applicationSettings>
 	<system.web>
 		<authentication mode="None" />
+		<globalization culture="en-AU" uiCulture="en-AU" />
```
Comment (s3364859) [inline CPT331.Web/Web.config]: Date formats were coming out backwards!

File: CPT331.Web/Controllers/AdminController.cs [hunk 15]
```diff
@@ -191,54 +190,65 @@ public ActionResult Crimes(string sortBy, SortDirection? sortDirection, uint? pa
...
-        /// <param name="id">The unique ID for the LocalGovernmentArea record.</param>
-        /// <returns>The 'Edit LocalGovernmentArea' view.</returns>
-        /// <permission cref="AdminAuthorizeAttribute">Only authorised personnel have access to this method.</permission>
+        /// <returns>LocalGovernmentAreas.csv</returns>
         [HttpGet]
-        public ActionResult LocalGovernmentArea(uint id)
+		public ActionResult ExportLocalGovernmentAreas()
+
+		{
+			return new ExportDataActionResult("LocalGovernmentAreas.csv", DataProvider.LocalGovernmentAreaRepository);
```
Comment (6nop) [inline CPT331.Web/Controllers/AdminController.cs]: Perhaps not the most useful information without the area column.

Maybe we should use the area.STText() in the proc after all?

File: CPT331.Data/Repository.cs (hunk 9 above)
Comment (6nop) [inline CPT331.Data/Repository.cs]: Slow, yes... but very reusable. I'm happy with that.

The CSV export will only get used occasionally anyway.

File: CPT331.Web/Web.config (hunk 14 above)
Comment (6nop) [inline CPT331.Web/Web.config]: Bloody yanks!

**Comment by 6nop** [comment]
Not an issue with this PR, but we'll have to remember to implement the export into the Crime table.

**Comment by s3364859** [comment]
@6nop, that's coming in the next PR :-)

File: CPT331.Core/ObjectModel/ReadOnlyDataObject.cs (hunk 2 above)
Comment (6nop) [inline CPT331.Core/ObjectModel/ReadOnlyDataObject.cs]: I just noticed an error in the comment:

**fron** a data source.

**Pull request merged**